pkg/archive/archive.go

```diff
@@ -1143,6 +1144,35 @@ loop:
 	return nil
 }
 
+// createImpliedDirectories will create all parent directories of the current path with default permissions, if they do
+// not already exist. This is possible as the tar format supports 'implicit' directories, where their existence is
+// defined by the paths of files in the tar, but there are no header entries for the directories themselves, and thus
+// we most both create them and choose metadata like permissions.
+//
+// The caller should have performed filepath.Clean(hdr.Name), so hdr.Name will now be in the filepath format for the OS
+// on which the daemon is running. This precondition is required because this function assumes a OS-specific path
+// separator when checking that a path is not the root.
+func createImpliedDirectories(dest string, hdr *tar.Header, options *TarOptions) error {
+	// Not the root directory, ensure that the parent directory exists
+	if !strings.HasSuffix(hdr.Name, string(os.PathSeparator)) {
```

Comment:
While looking at this, I was testing some variations, and found an issue with `idtools.MkdirAllAndChownNew`; opened a PR with a possible fix for that; https://github.com/moby/moby/pull/44203

Reply:
Yes, I agree with all of that. This was mostly me trying to understand what it does (it's fun how much you can spend on a very simple, single line of code! 😂).

I thought I'd write it down, in case there's something we overlooked, and for my "future self".

Reply:
I'll put this one in my notes as a spot to revist when I have some time. Certainly, a holistic look at this code with fresh eyes would be useful, given its foundational nature and possible security implications.